### PR TITLE
Make Response object return optional if such config selected

### LINKED ISSUE #290: Introduce Optionals
We have had 2 bugs and contributions due to NPE somewhere we did not expect. I'd like to add Optional to the project.  I have no preference to our own smaller `Optional` type or to bring in a library that does it.  I can't imagine this being more than 20 methods.

### PR DESCRIPTION
**Pull request by sav007**
Closes #290 

Address issue with `Response` data field doesn't return data wrapped with optional.

Take a look at https://github.com/apollographql/apollo-android/pull/331/files#diff-bd1f7dd25c2f308ad46f434f16893f11 and https://github.com/apollographql/apollo-android/pull/331/files#diff-1d4349ba6464dd54d277f125b3253316 as example how it will be used.

### CODE REVIEW

File: apollo-api/src/main/java/com/apollographql/android/api/graphql/Operation.java [hunk 1]
```diff
@@ -6,14 +6,16 @@
...
   /** TODO */
   String queryDocument();
 
   /** TODO */
   V variables();
 
-  ResponseFieldMapper<? extends Operation.Data> responseFieldMapper();
+  ResponseFieldMapper<D> responseFieldMapper();
+
+  T wrapData(D data);
```
Comment (sav007) [inline apollo-api/src/main/java/com/apollographql/android/api/graphql/Operation.java]: will be called by ResponseConverter

File: apollo-compiler/src/test/graphql/com/example/arguments_complex/TestQuery.java [hunk 2]
```diff
@@ -44,12 +44,17 @@ public String queryDocument() {
   }
 
   @Override
+  public Optional<TestQuery.Data> wrapData(TestQuery.Data data) {
+    return Optional.fromNullable(data);
```
Comment (sav007) [inline apollo-compiler/src/test/graphql/com/example/arguments_complex/TestQuery.java]: Example how data will be wrapped with optional

File: apollo-runtime/src/main/java/com/apollographql/android/impl/HttpResponseBodyConverter.java [hunk 3]
```diff
@@ -55,7 +55,7 @@
         }
       }
       jsonReader.endObject();
-      return new Response<>(operation, data, errors);
+      return new Response<>(operation, operation.wrapData(data), errors);
```
Comment (sav007) [inline apollo-runtime/src/main/java/com/apollographql/android/impl/HttpResponseBodyConverter.java]: this call will wrap data

**Comment by BenSchwab** [comment]
Do we have any tests that test what happens when `optional` is disabled?

**Comment by marwanad** [comment]
@BenSchwab There aren't any for the plugin extension but there are for the [compiler generation](https://github.com/apollographql/apollo-android/pull/322/files#diff-be5ed569a931b8335d61a5097b179cccR77)

**Comment by sav007** [comment]
Yes, that correct. We can have 2 Optional implementations: guava or our own shadowed version. So boolean field is not enough. Plus our Runtime lib knows nothing about guava. So only if user uses Guava we will wrap data (via method `wrapData`) to guava Optional, otherwise we will wrap it into our shadowed Optional

**Comment by sav007** [comment]
@digitalbuddha @marwanad any comments?

**Comment by digitalbuddha** [comment]
Not from first glance need to look tonight

**Pull request merged**